cmd/main.go

```diff
@@ -2,26 +2,139 @@ package main
 
 import (
 	"log"
+	"net/http"
 	"os"
+	"time"
+
+	"github.com/PDOK/azure-storage-usage-exporter/internal/du"
+
+	"github.com/google/uuid"
+
+	"github.com/PDOK/azure-storage-usage-exporter/internal/serv"
+	"github.com/go-co-op/gocron/v2"
+	"github.com/iancoleman/strcase"
+	"github.com/prometheus/client_golang/prometheus/promhttp"
+
+	"github.com/PDOK/azure-storage-usage-exporter/internal/agg"
+
+	"gopkg.in/yaml.v2"
 
 	"github.com/urfave/cli/v2"
 )
 
+type Config struct {
+	Labels agg.Labels            `yaml:"labels"`
+	Rules  []agg.AggregationRule `yaml:"rules"`
+}
+
+const (
+	cliOptAzureStorageConnectionString = "azure-storage-connection-string"
+	cliOptBindAddress                  = "bind-address"
+	cliOptBlobInventoryContainer       = "blob-inventory-container"
+	cliOptConfigFile                   = "config"
+)
+
 var (
-	cliFlags = []cli.Flag{}
+	cliFlags = []cli.Flag{
+		&cli.StringFlag{
+			Name:     cliOptAzureStorageConnectionString,
+			Usage:    "Connection string for connecting to the Azure blob storage that holds the inventory",
+			Required: true,
+			EnvVars:  []string{strcase.ToScreamingSnake(cliOptAzureStorageConnectionString)},
+		},
+		&cli.StringFlag{
+			Name:    cliOptBindAddress,
+			Usage:   "The TCP network address addr that is listened on.",
+			Value:   ":8080",
+			EnvVars: []string{strcase.ToScreamingSnake(cliOptBindAddress)},
+		},
+		&cli.StringFlag{
+			Name:    cliOptBlobInventoryContainer,
+			Usage:   "Name of the container that holds the inventory",
+			Value:   "blob-inventory",
+			EnvVars: []string{strcase.ToScreamingSnake(cliOptBlobInventoryContainer)},
+		},
+		&cli.StringFlag{
+			Name:      cliOptConfigFile,
+			Usage:     "Config file with aggregation labels and rules",
+			EnvVars:   []string{strcase.ToScreamingSnake(cliOptConfigFile)},
+			Required:  true,
+			TakesFile: true,
+		},
+	}
 )
 
 func main() {
 	app := cli.NewApp()
-	app.Name = "PDOK Storage Usage Explorer"
+	app.HelpName = "Azure Storage Usage Exporter"
+	app.Name = "azure-storage-usage-exporter"
+	app.Usage = "Aggregates an Azure Blob Inventory Report and export as Prometheus metrics"
 	app.Flags = cliFlags
-	app.Action = func(_ *cli.Context) error {
-		log.Println("hello world")
-		return nil
+	app.Action = func(c *cli.Context) error {
+		aggregator, err := createAggregatorFromCliCtx(c)
+		if err != nil {
+			return err
+		}
+		metricsUpdater := serv.NewMetricsUpdater(aggregator, "pdok", "storage", 1000)
+
+		scheduler, err := gocron.NewScheduler()
+		if err != nil {
+			return err
+		}
+		_, err = scheduler.NewJob(
```

Comment:
Zelf actief data - op basis van scheduling - ophalen is denk ik niet optimaal. Het is sowieso niet idiomatic voor een Prometheus exporter: https://prometheus.io/docs/instrumenting/writing_exporters/#scheduling. Punt is dat je nu de interval op 2 punten moet configureren: hier in de exporter, en daarnaast nog de scrape interval in Prometheus zelf. 

Idee is om het alleen in Prometheus te doen, via de scrape interval. De data kan je dan wel cachen in de exporter. Je zou prima een cache ttl van bijv 4 uur kunnen hanteren. Dan komt er nog minder load/kosten op Azure Blob. 

Flow zou er dan zou uitzien:
- 1x bij startup de cache populaten.
- bij elke scrape uit cache serveren
- indien ttl verstreken: cache opnieuw populaten.

Reply:
Je hoeft ook niet in de scope van het scrape request alle data te verzamelen. Je zou prima tijdens het scrape verzoek async de data kunnen verzamelen. Dus in een losse goroutine. Elke scrape verzoek = data uit cache serveren, en soms - daarnaast - een nieuwe inventory report read starten.

Reply:
ok, als dat de idiomatic way is ga ik dat doen. ik wou het nog vandaag doen maar ik kwam er niet aan toe. ik wil hem eerst even in k8s hangen. en dan refactor ik dit. dus ik merge hem wel vast.